File: 7.Go版每日一题/787. K 站中转内最便宜的航班.ts
function findCheapestPrice1(n: number, flights: number[][], src: number, dst: number, k: number): number {
	const INF = 10000 * 101 + 1
	const f = new Array(k + 2).fill(0).map(() => new Array(n).fill(INF))
	f[0][src] = 0
	for (let t = 1; t <= k + 1; ++t) {
		for (const flight of flights) {
			const j = flight[0],
				i = flight[1],
				cost = flight[2]
			f[t][i] = Math.min(f[t][i], f[t - 1][j] + cost)
		}
	}
	let ans = INF
	for (let t = 1; t <= k + 1; ++t) {
		ans = Math.min(ans, f[t][dst])
	}
	return ans == INF ? -1 : ans
}

function findCheapestPrice(n: number, flights: number[][], src: number, dst: number, k: number): number {
	const INF = 10000 * 101 + 1
	let f = new Array(n).fill(INF)
	f[src] = 0
	let ans = INF
	for (let t = 1; t <= k + 1; ++t) {
		const g = new Array(n).fill(INF)
		for (const flight of flights) {
			const j = flight[0],
				i = flight[1],
				cost = flight[2]
			g[i] = Math.min(g[i], f[j] + cost)
		}
		f = g
		ans = Math.min(ans, f[dst])
	}
	return ans == INF ? -1 : ans
}

File: 7.Go版每日一题/787. K 站中转内最便宜的航班.go
package kStation

func main() {}

func findCheapestPrice(n int, flights [][]int, src int, dst int, k int) int {
	const inf = 10000*101 + 1
	f := make([]int, n)
	for i := range f {
		f[i] = inf
	}
	f[src] = 0
	ans := inf
	for t := 1; t <= k+1; t++ {
		g := make([]int, n)
		for i := range g {
			g[i] = inf
		}
		for _, flight := range flights {
			from, to, cost := flight[0], flight[1], flight[2]
			g[to] = min(g[to], f[from]+cost)
		}
		f = g
		ans = min(ans, f[dst])
	}
	if ans == inf {
		ans = -1
	}
	return ans
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

/*

func findCheapestPrice(n int, flights [][]int, src int, dst int, k int) int {
	const inf = 10000*101 + 1
	f := make([][]int, k+2)
	for i := range f {
			f[i] = make([]int, n)
			for j := range f[i] {
					f[i][j] = inf
			}
	}
	f[0][src] = 0
	for t := 1; t <= k+1; t++ {
			for _, flight := range flights {
					j, i, cost := flight[0], flight[1], flight[2]
					f[t][i] = min(f[t][i], f[t-1][j]+cost)
			}
	}
	ans := inf
	for t := 1; t <= k+1; t++ {
			ans = min(ans, f[t][dst])
	}
	if ans == inf {
			ans = -1
	}
	return ans
}

func min(a, b int) int {
	if a < b {
			return a
	}
	return b
}
*/
